### PR TITLE
Incorporate kernel 5.15 api changes

### PR DESCRIPTION
**Pull request by prabirpaul**
**What this PR does / why we need it**:
Px-fuse changes for building and running on kernel >= 5.15 

**Which issue(s) this PR fixes** (optional)
Closes #
PWX-21857

**Special notes for your reviewer**:

### CODE REVIEW

**Review by jrivera-px — APPROVED** [review]

File: pxd.c [hunk 1]
```diff
@@ -1134,24 +1134,13 @@ static int pxd_init_disk(struct pxd_device *pxd_dev, struct pxd_add_ext_out *add
...
-		return -ENOMEM;
-
-	snprintf(disk->disk_name, sizeof(disk->disk_name),
-			PXD_DEV"%llu", pxd_dev->dev_id);
-	disk->major = pxd_dev->major;
-	disk->first_minor = pxd_dev->minor;
-	disk->flags |= GENHD_FL_EXT_DEVT | GENHD_FL_NO_PART_SCAN;
-	disk->fops = &pxd_bd_ops;
-	disk->private_data = pxd_dev;
-
```
Comment (sulakshm) [inline pxd.c:1154]: given disk is not initialized, below L1152 wherein disk->fops will fail, if pxd is compiled with MAKEREQ macro (old ways of doing block IO). Actually take the above block under 
#ifdef __PXD_BIO_MAKEREQ_
/// alloc_disk L1142->L1153 above
#endif

We only support PXD_BIO_BLKMQ always.

File: pxd.c [hunk 2]
```diff
@@ -1173,8 +1162,7 @@ static int pxd_init_disk(struct pxd_device *pxd_dev, struct pxd_add_ext_out *add
 		q = blk_alloc_queue(GFP_KERNEL);
 #endif
 		if (!q) {
-			err = -ENOMEM;
-			goto out_disk;
+			return -ENOMEM;
```
Comment (sulakshm) [inline pxd.c:1165]: then here put_disk is needed. This block is under PXD_BIO_MAKEREQ.

File: pxd.c [hunk 3]
```diff
@@ -1197,24 +1185,54 @@ static int pxd_init_disk(struct pxd_device *pxd_dev, struct pxd_add_ext_out *add
...
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5,15,0)
+	  disk = blk_mq_alloc_disk(&pxd_dev->tag_set, pxd_dev);
+	  if (IS_ERR(disk)) {
+		blk_mq_free_tag_set(&pxd_dev->tag_set);
+		return PTR_ERR(disk);
+	  }
+	  q = disk->queue;
+#else
+	  /* Create gendisk info. */
```
Comment (sulakshm) [inline pxd.c:1200]: alloc_disk is called in both ifdef and ifndef PX_BLKMQ... can be taken out and if kernel version < 5,15,0 then can perform alloc_disk. outside above L1192.

File: pxd.c [hunk 4]
```diff
@@ -1197,24 +1185,54 @@ static int pxd_init_disk(struct pxd_device *pxd_dev, struct pxd_add_ext_out *add
...
 #endif /* __PX_BLKMQ__ */
 #endif /* __PXD_BIO_MAKEREQ__ */
 
+	snprintf(disk->disk_name, sizeof(disk->disk_name),
+		 PXD_DEV"%llu", pxd_dev->dev_id);
+	disk->major = pxd_dev->major;
+	disk->minors = 256;
+	disk->first_minor = pxd_dev->minor;
+	disk->flags |= GENHD_FL_EXT_DEVT | GENHD_FL_NO_PART_SCAN;
+	disk->fops = &pxd_bd_ops;
```
Comment (sulakshm) [inline pxd.c:1232]: The above init is incorrect if MAKEREQ is used. need following code.
#ifdef __PXD_BIO_MAKEREQ__
		pxd_printk("adding disk as makereq device %llu", pxd_dev->dev_id);
		disk->fops = get_bd_fpops();
#endif

File: pxd.c [hunk 5]
```diff
@@ -1465,7 +1483,11 @@ ssize_t pxd_export(struct fuse_conn *fc, uint64_t dev_id)
 	struct pxd_device *pxd_dev = find_pxd_device(ctx, dev_id);
 
 	if (pxd_dev) {
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5,15,0)
+		device_add_disk(&pxd_dev->dev, pxd_dev->disk, NULL);
+#else
 		add_disk(pxd_dev->disk);
+#endif
```
Comment (sulakshm) [inline pxd.c:1490]: ok

File: pxd.c [hunk 6]
```diff
@@ -1519,13 +1541,21 @@ ssize_t pxd_remove(struct fuse_conn *fc, struct pxd_remove_out *remove)
 		QUEUE_FLAG_SET(QUEUE_FLAG_DYING, pxd_dev->disk->queue);
 
 		mutex_unlock(&pxd_dev->disk->queue->sysfs_lock);
+#else
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5,15,0)
+		// Do not mark queue dead, del_gendisk will try to
+		// submit all outstanding IOs on this device
 #else
 		blk_set_queue_dying(pxd_dev->disk->queue);
+#endif
```
Comment (sulakshm) [inline pxd.c:1550]: ok

File: pxd.c [hunk 7]
```diff
@@ -1519,13 +1541,21 @@ ssize_t pxd_remove(struct fuse_conn *fc, struct pxd_remove_out *remove)
...
 #endif
 	}
 
 	spin_unlock(&pxd_dev->lock);
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5,15,0)
+	pxd_free_disk(pxd_dev);
+#endif
```
Comment (sulakshm) [inline pxd.c:1558]: ok

File: pxd.c [hunk 8]
```diff
@@ -1270,15 +1283,20 @@ static void pxd_free_disk(struct pxd_device *pxd_dev)
...
-	if (disk->flags & GENHD_FL_UP) {
+	if (disk) {
 		del_gendisk(disk);
-		if (disk->queue)
-			blk_cleanup_queue(disk->queue);
+
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5,15,0)
+		blk_cleanup_disk(disk);
+#else
+		blk_cleanup_queue(disk->queue);
```
Comment (sulakshm) [inline pxd.c:1292]: if (disk->queue != NULL) missing before calling blk_cleanup_queue

File: pxd.c [hunk 9]
```diff
@@ -1270,15 +1283,20 @@ static void pxd_free_disk(struct pxd_device *pxd_dev)
...
 	pxd_fastpath_cleanup(pxd_dev);
 	pxd_dev->disk = NULL;
-	if (disk->flags & GENHD_FL_UP) {
+	if (disk) {
 		del_gendisk(disk);
-		if (disk->queue)
-			blk_cleanup_queue(disk->queue);
+
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5,15,0)
+		blk_cleanup_disk(disk);
```
Comment (sulakshm) [inline pxd.c:1290]: again call this under disk->queue != NULL

**Review by sulakshm — COMMENTED** [review]
minor nits inline, mostly okay.

**Comment by sulakshm** [comment]
also please check both normal/fastpath volume attach/detach happens okay under good scenario and px restarts.

**Review by sulakshm — APPROVED** [review]
Looks good. 
Please do run one more unit tests especially create/attach/detach/delete cycles in both native and fastpath specific volumes.

**Comment by prabirpaul** [comment]
```
Test1:
for((i=1;i<100;i++)); do pxctl v c --nodes LocalNode --fastpath v$i; pxctl host attach v$i; pxctl host detach v$i; pxctl v d v$i -f; done
for((i=1;i<100;i++)); do pxctl v c --nodes LocalNode v$i; pxctl host attach v$i; pxctl host detach v$i; pxctl v d v$i -f; done

5.15.13-1.el7 btrfs, 5.15.13-1.el7 dmthin
5.7.12-1.el7 dmthin

Test2:
- Create and attach 10 volumes, stop px, delete all volume from another node, restart px
- Make sure force detach goes through, px comes back up
 
5.15.13-1.el7 dmthin
5.7.12-1.el7 dmthin
```

**Pull request merged**